### ISSUE TITLE
Virtual File-system Memory Leak

### ISSUE THREAD

**Issue by sulljason** [MightyPirates/OpenComputers#1411]
The heap usage slowly climbs until it runs out of memory and thrashes the garbage collector. Object arrays of hashmaps which are in the top 3 memory users just below the byte arrays for the world. I took a heap dump and looked at the keys of the hashmaps and their values were things like "buffer.lua" and "keyboard.lua".

Heap dump: http://hydra.13-thirtyseven.com/Private/heapdump-1441762677841.7z

VisualVM:
![screenshot](https://cloud.githubusercontent.com/assets/1346094/9751935/a7fe0814-565f-11e5-8c07-42078b846a25.png)

![screenshot-1](https://cloud.githubusercontent.com/assets/1346094/9751970/0c83db24-5660-11e5-87e2-daec22cb5f4e.png)

![screenshot-2](https://cloud.githubusercontent.com/assets/1346094/9751977/1eede5c0-5660-11e5-9553-f75363a1a942.png)

EDIT:
Downgrading to OpenComputers-MC1.7.10-1.5.16.32-universal seems to have fixed the issue. Server has been up for 9h or so and the heap is looking stable. (That random GC spike was a manual collection.)
![screenshot](https://cloud.githubusercontent.com/assets/1346094/9764877/4539d41c-56c6-11e5-98ba-e634ae2f9958.png)

**Comment by 1n5aN1aC** [comment]
This is OpenComputers MC1.7.10-1.5.17.33  by the way.

We updated 3 mods last time we updated, and this started happening:
1.  Open Computers
2.  Botania
3.  StorageDrawers

Searching for more proof that this is OpenComputers, and EDIT:  Heap dump now included in Sulljason's initial post.

**Comment by fnuecke** [comment]
Thanks for the heap dump, will have a look! It's odd that downgrading to 1.5.16 should change anything, though, since [nothing filesystem related changed](https://github.com/MightyPirates/OpenComputers/compare/v1.5.16...v1.5.17) (barely anything changed, really).

**Comment by 1n5aN1aC** [comment]
Yeah, I was a bit confused about that as well.  I haven't spent much time digging through the Heap dump, but @sulljason has looked through it, and thinks it's definitely file system related.

All that I know for sure is that I pushed an update for our private modpack, just rolling back OC to "MC1.7.10-1.5.16.32" seems to have very much fixed it.

So I really have no clue what could be causing this.  :(

**Comment by fnuecke** [comment]
So I've dug through the heap dump a bit, and from what I can see, all those nodes aren't really related to OC at all (based on randomly looking at a bunch [~50-100 I guess] of them). I've been trying to write an OQL query to get the GC roots for all the nodes, but I haven't used it enough to get it to work (in a timely manner :P), so if someone else wants to give that a shot...

**Comment by sulljason** [comment]
Is the problem with rummaging through 1000s of objects eh? Get completely different answers based on the ones you randomly select lol. Doesn't really explain why downgrading fixed the issue completely though.

**Comment by fnuecke** [comment]
Aye. But again, there's really nothing filesystem (or even hashmap I don't think) related that changed between those versions, so that's weird anyway :-/

**Comment by 1n5aN1aC** [comment]
After a bit of searching around, I discovered there was a recent update to logistics pipes that fixes an odd memory leak.

I'm going to try updating Logistics, as well as re-updating OpenComputers to the latest version, to see if the OpenComputers culprit was really just an odd coincidence.

**Comment by fnuecke** [comment]
I did notice some references to the VFSes coming through LP (via PipeSomethingSatelliteSomething IIRC?), but wasn't sure that'd be related. Would be great if that'd solve it.

**Comment by 1n5aN1aC** [comment]
Okay then, I didn't notice those, but the bug report I was reading definitely mentioned the fix being related to Satellite Crafting Pipes, so there's a very high probability that this update will fix the problem, and the open computers update just being a strange timing fluke.

If you want, you can close this issue, as it's quite likely that's what it is.  If not, We'll play a bit, see if it's fixed, and @sulljason or I'll come back and close it if it's fixed in a day or two.

**Comment by fnuecke** [comment]
So I take it updating Logistics Pipes fixed it? Good to hear :-)

**Comment by 1n5aN1aC** [comment]
@fnuecke Yup, seems that that was what the issue was.

I guess it coinciding with updating OC was just a co-incidence. :)

**Comment by sulljason** [comment]
But, but  we rolled back only OC and it fixed it and "We updated 3 mods last time we updated, and this started happening:
1. Open Computers
2. Botania
3. StorageDrawers"
logistics pipes isn't on the list.

**Comment by 1n5aN1aC** [comment]
You are correct, it really doesn't make sense, but I'm wondering if it co-coincided when we added satellite crafting functionality to our logistics system, then due to the fact that the bug is relatively elusive.  (Has to do with chunk loading / unloading weirdness)  it took a while to manifest itself.

At any rate, it hasn't reared it's ugly head again so far, so It definitely seems like something we did fixed it.  (Most likely that update.)
